### DISCUSSION TITLE
Should support for TensorFlow be dropped given its library anti-patterns?

### DISCUSSION THREAD

**Discussion by matthewfeickert** [scikit-hep/pyhf#1595] [Governance and Maintenance]
TensorFlow is continually becoming a burden to support and is arguably slowing down project development given library development anti-patterns like https://github.com/tensorflow/tensorflow/pull/40789 and https://github.com/tensorflow/tensorflow/issues/51743 where dependencies are being pinned down to the patch level. At this point, TensorFlow is **not** a Python library &mdash; it is a Python application that you have to build your project around. That's fine, but it means that we can't effectively treat `pyhf` as a Python library then if we want to provide support across all backends.

I see two ways forward:
1. Drop all TensorFlow support entirely with `pyhf` release `v0.7.0`.
2.
   - Figure out how to test backends independently in the tests (c.f. Issue #1454)
   - Remove TensorFlow from the `backends` extra and keep it as a `tensorflow` extra
   - Test the `tensorflow` extra in a fully different workflow and say that we will support a particular minor release of TensorFlow and TensorFlow Probability for a particular release cycle of `pyhf`.
   
TensorFlow is already the **slowest** backend that `pyhf` has and it is probably because we haven't optimized `pyhf` around its graph options. My guess is that no on using `pyhf` is actually using the TensorFlow backend anyway and is using JAX instead. I'm not sure that given the goals of use that `pyhf` has in terms of particle physics if there is any advantage to continuing support for TensorFlow. My inclination is to drop it, but I'm curious what others think.

cc @lukasheinrich @kratsg @alexander-held @jpivarski @henryiii @eduardo-rodrigues

**Comment by jpivarski** [comment]
I don't have direct experience with TensorFlow, but if what you say is true, that's a good thing to know and to be wary of in other HEP projects.

I remember that [zfit](https://github.com/zfit/zfit) once relied heavily on TensorFlow, but I don't know if it still does. @jonas-eschle, what is your take on the usability of TensorFlow _as a library_, for HEP? Is it getting more application-like than library-like?

**Reply by matthewfeickert** [reply]
TensorFlow is quite aggressive as _all_ of its dependencies in `setup.py` use the compatible release syntax. For example, this is taken from its [`setup.py` for TensorFlow `v2.6.0`](https://github.com/tensorflow/tensorflow/blob/919f693420e35d00c8d0a42100837ae3718f7927/tensorflow/tools/pip_package/setup.py#L77-L107)

```python
REQUIRED_PACKAGES = [
    # NOTE: As numpy has releases that break semver guarantees and several other
    # deps depend on numpy without an upper bound, we must install numpy before
    # everything else.
    'numpy ~= 1.19.2',
    # Install other dependencies
    'absl-py ~= 0.10',
    'astunparse ~= 1.6.3',
    'clang ~= 5.0',
    'flatbuffers ~= 1.12.0',
    'google_pasta ~= 0.2',
    'h5py ~= 3.1.0',
    'keras_preprocessing ~= 1.1.2',
    'opt_einsum ~= 3.3.0',
    'protobuf >= 3.9.2',
    'six ~= 1.15.0',
    'termcolor ~= 1.1.0',
    'typing_extensions ~= 3.7.4',
    'wheel ~= 0.35',
    'wrapt ~= 1.12.1',
    # These packages need to be pinned exactly as newer versions are
    # incompatible with the rest of the ecosystem
    'gast == 0.4.0',
    # TensorFlow ecosystem packages that TF exposes API for
    # These need to be in sync with the existing TF version
    # They are updated during the release process
    # When updating these, please also update the nightly versions below
    'tensorboard ~= 2.6',
    'tensorflow_estimator ~= 2.6',
    'keras ~= 2.6',
]
```

I'm not concerned about the `tensorboard`, `tensorflow_estimator`, and `keras` restrictions, but is isn't very nice that it makes such tight constraints on `numpy`, `h5py`, `opt_einsum`, and `typing_extensions`.

TensorFlow Probability is much better (here is [`v0.13.0`](https://github.com/tensorflow/probability/blob/9e50fa39f37216dd084c24aa3dd14ab0d4dee926/setup.py#L30-L37))
```python
REQUIRED_PACKAGES = [
    'six >= 1.10.0',
    'numpy >= 1.13.3',
    'decorator',
    'cloudpickle >= 1.3',
    'gast >= 0.3.2',  # For autobatching
    'dm-tree'  # For NumPy/JAX backends (hence, also for prefer_static)
]
```

but unless you're using it with JAX then you need TensorFlow.

> I remember that [zfit](https://github.com/zfit/zfit) once relied heavily on TensorFlow, but I don't know if it still does.

`zfit` is still TensorFlow only. c.f. their [`requirements.txt` which is read by `setup.py`](https://github.com/zfit/zfit/blob/6cc0a5cadaa9f13c1e6e44b093b4f38658e072fb/requirements.txt)

```python
colorama
colored
colorlog
dotmap
iminuit>=2.3
ipyopt<0.12
nlopt
numdifftools
numpy>=1.16
ordered-set
pandas
scipy>=1.2
tabulate
tensorflow>=2.5,<2.7
tensorflow-addons
tensorflow_probability>=0.12,<0.14
texttable
tf_quant_finance
uproot>=4,<5
```

**Reply by matthewfeickert** [reply]
For comparison:

## `jax` and `jaxlib`

[`jax` `v0.2.20`](https://github.com/google/jax/blob/a7b61c0e00d1b535df8a30a82edc0074884d5f4c/setup.py#L38-L43)
```python
    install_requires=[
        'absl-py',
        'numpy>=1.18',
        'opt_einsum',
        'scipy>=1.2.1',
    ],
```

[`jaxlib` `v0.1.71`](https://github.com/google/jax/blob/63a7ad9ec334d31b23c8e86f16408972429c8682/jaxlib/setup.py#L35)
```python
    install_requires=['scipy', 'numpy>=1.18', 'absl-py', 'flatbuffers >= 1.12, < 3.0'],
```

## PyTorch

[`torch` `v1.8.2`](https://github.com/pytorch/pytorch/blob/e0495a7aa104471d95dc85a1b8f6473fbcc427a8/setup.py#L355-L358)
```python
install_requires = [
    'typing_extensions',
    'dataclasses; python_version < "3.7"'
]
```

**Comment by alexander-held** [comment]
A current example for the consequences of the restrictive dependency pinning in `tensorflow` is the following:
```bash
pip install black tensorflow tensorflow-probability
```
will first install `tensorflow` 2.6.0, and then downgrade to 2.3.4 in a few steps due to a `typing-extensions` conflict. `black` requires `typing-extensions>=3.10.0.0` (as of `black` 21.8b0), while `tensorflow` requires `typing-extensions~=3.7.4`. In the end this installation breaks `tensorflow-probability`:
```python
import tensorflow, tensorflow_probability
```
will fail with

> ImportError: This version of TensorFlow Probability requires TensorFlow version >= 2.5; Detected an installation of version 2.3.4. Please upgrade TensorFlow to proceed.

This combination of packages comes up for example with `pip install black pyhf[tensorflow]`, and also appears when updating `cabinetry` to `pyhf` 0.6.3 (https://github.com/alexander-held/cabinetry/pull/248).

**Reply by lukasheinrich** [reply]
I would vote for keeping it but having it be it's own extra

**Reply by alexander-held** [reply]
From the `cabinetry` side, I am considering dropping official support for the TF backend of `pyhf`. Dependency resolution for the `cabinetry` dev environment in Python 3.9 (including `pyhf[backends]`) is extremely slow, I stopped it after half an hour of CI. It should still work fine if users manage to install it correctly themselves, but would not be tested any more as it looks like it adds a significant amount of overhead.

**Reply by matthewfeickert** [reply]
> I would vote for keeping it but having it be it's own extra

Yeah, after all the work that we've put in just to get it working it feels dissatisfying to just drop all support for it. Though the level of restriction that it makes `pip`'s dependency resolver put on everything else at install time is becoming unfeasible.

> From the `cabinetry` side, I am considering dropping official support for the TF backend of `pyhf`.

@alexander-held I don't think that anyone on the `pyhf` side would fault `cabinetry` for dropping support for one of our backends. The maintenance costs should be decided at each project.

**Reply by alexander-held** [reply]
Following a suggestion from @matthewfeickert, we can keep all backends supported in `cabinetry` by moving them to a separate setup extra that is installed in a second step. This results in a usable environment and makes the dependency resolution much faster too.

**Comment by jonas-eschle** [comment]
Hey, that is indeed a good discussion and a thought worth. We do in zfit still mainly rely on TensorFlow, however, it is also compatible with other libraries (_because_ we use TF and not e.g. JAX). Let me elaborate, it's a somewhat love-hate relationship with TF having a few (annoying) drawbacks but, AFAIU, TF still being *the most powerful library* for computations in HEP.

First, comparing to alternatives, we can neglect numba (as having a quite different concept of compiling, not tracing) and looking at jit-traceable, autograd supporting libraries (JAX, torch,...).

About the performance and graph: I had a brief look at it and it seems (AFAI see) that pyhf does not really jit TF? Since TF 2 the computational model completely changed. It's not a graph anymore (technically, but implementation detail) but a jittable library like JAX. Like jax, it actually supports XLA. I think in pyhf this is always eager, no jitting, and also no xla therefore (`jit_compile` flag in `tf.function`).

I did a preliminary test using the benchmarks and TF is exactly at the same speed as JAX and torch.
I can take a look at this if you're interested, in general TF is much like JAX (and that's how we should also wrap the objective function). *Update*: to be precise, TF is in my benchmarks even up to 5-10% faster than others.

Now for TensorFlow:

Disadvantages:
- Yes, dependencies. As you mentioned about the patch release pin down, I think it's just pinned to the minor release (`~=` the wiggle leaves the last dimension floating). That is still a heavy requirement for a library. The main trouble to me comes currently with with the requirement of `numpy<1.20`. TF is rather slow in adopting this things, however I would see it still quite as a library, not an application, just a rather slow moving one.
- packaging-wise: there are a few additional libraries (using TF) that provide utilities that are not on conda


Mixed:
- the API is different from numpy, yet they offer a (still experimental) [numpy API](https://www.tensorflow.org/api_docs/python/tf/experimental/numpy?version=nightly). Apart from that, I find their API in parts even superior to Numpy in terms of name consistency (for example there is a `stack` and `unstack`, the `shape` is always called `shape`, not sometimes `size`)

Advantages, mainly compared to JAX:
- TensorFlow can handle unknown shapes jitted. JAX for example (or XLA in general), can't. This is a large thing for an open-world fitter like zfit because whenever you do something dynamically, you have unknown shapes. Accept-reject sampling, adaptive integration.... It happens more than I thought and restricting is a quite limiting thing. **For pyhf** this is not a problem I think as shapes are always known (if it works with JAX that is already confirmation that it does).
- Interoperability: When I looked at swapping the backend the last time, this is what I remember. TensorFlow _does_ stick to the numpy array interface. As the only library AFAIK. You can give it anything that has the array protocol implemented (numpy, but also jax, torch...), which [JAX can't handle](https://github.com/google/jax/issues/6222)
- at least currently, JAX still lacks a lot of content (such as [no float64 support for convolutions](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html?highlight=convolutions%2064%20bite#caveats). However, this is not fundamental and will (presumably) improve over time, at some point I assume that JAX will be superior in content.

Advantages, mainly compared to PyTorch:
- TensorFlow has AFAIK way better support for more mathematical functionality such as complex, jagged, sparse.

**Conclusion TF for HEP**

TF is not the perfect library for HEP, indeed. But it seems to me still the most capable library around, with advantages (such as unknown shape handling) that are crucial. I hope and assume that JAX one day will be superior. The current disadvantages are mostly of package technical nature. Given also it's recent switch to a completely new model, the whole package seems to be stabilizing. But yes, it's big and feels somewhat intrusive.

This is why I think it's still very capable _for HEP in general, including zfit_.

**TF for pyhf**

The packaging can really increase the burden for no reason. I think that a swappable backend is _nice_, but more of an implementation detail (it means a library needs to restrict itself to the subset of all common features, effectively removing features). The more important aspect to me seems to be interoperable with the rest. And as pyhf has the advantage of needing only this subset, having static shapes, I think this is very reasonable.

I would suggest to go along the lines @matthewfeickert proposed:
- give it a try with jitting and see the speed gain
- since the problems are related to dependencies basically: drop the "support" as long as it does, unfortunately, not behave relaxed enough with requirements and is a problem, remove it from the default backends, run the normal ci without it and (maybe) have a runner with TF that is allowed to fail. In this way we see once we break something. If TF gets better with dependencies, we can always add it back.

*So summarized*, I think TF is still the most capable library around for speedy computations, but big and slow moving. I would keep it in general, but not at the cost of not being able to install other packages

**Reply by jonas-eschle** [reply]
Just to add to my own post, while TF seems still very capable, there are more serious cracks with new features. For example https://github.com/tensorflow/tensorflow/issues/57365 is quite serious IMHO, but not addressed since over a year. It completely broke the zfit internals.
And of course, my all-time favorite of a serious GPU memory leak https://github.com/tensorflow/tensorflow/issues/36465#issuecomment-582749350 that we cannot simply "fix" by using processes.

**Comment by mihaimaruseac** [comment]
Currently working on untentangling this mess but progress is unfortunately slow (devinfra team severely reduced, cleaning up the mess not being considered a promotion worthy project, so very few takers for these; then issues with internal Google monorepo + single versioning policy)

**Comment by mihaimaruseac** [comment]
TF should now have a better setup.py

**Reply by matthewfeickert** [reply]
Yup. I've been following https://github.com/tensorflow/tensorflow/issues/51743#issuecomment-933317517 so was glad to see that today.

Nice work to you and your team @mihaimaruseac, as indeed, [this is quite an improvement](https://github.com/tensorflow/tensorflow/blob/3712a2d3455e6ccb924daa5724a3652a86f6b585/tensorflow/tools/pip_package/setup.py#L73-L104) and will be nice to have in the next release. :) 

```python
REQUIRED_PACKAGES = [
    # NOTE: As numpy has releases that break semver guarantees and several other
    # deps depend on numpy without an upper bound, we must install numpy before
    # everything else.
    'numpy >= 1.14.5', # keras 2.6 needs 1.19.2, h5py needs 1.14.5 for py37
    # Install other dependencies
    'absl-py >= 0.4.0',
    'astunparse >= 1.6.0',
    'libclang >= 9.0.1',
    'flatbuffers >= 1.12, < 3.0', # capped as jax 0.1.71 needs < 3.0
    'google_pasta >= 0.1.1',
    'h5py >= 2.9.0', # capped since 3.3.0 lacks py3.6
    'keras_preprocessing >= 1.1.1', # 1.1.0 needs tensorflow==1.7
    'opt_einsum >= 2.3.2', # sphinx pin not removed up til 3.3.0 release
    'protobuf >= 3.9.2',
    'six >= 1.12.0',
    'termcolor >= 1.1.0',
    'typing_extensions >= 3.6.6',
    'wheel >= 0.32.0, < 1.0', # capped as astunparse 1.6.0-1.6.3 requires < 1.0
    'wrapt >= 1.11.0',
    # These packages need to be pinned exactly as newer versions are
    # incompatible with the rest of the ecosystem
    'gast >= 0.2.1, < 0.5.0', # TODO(lpak): if this breaks, revert to 0.4.0
    # TensorFlow ecosystem packages that TF exposes API for
    # These need to be in sync with the existing TF version
    # They are updated during the release process
    # When updating these, please also update the nightly versions below
    'tensorboard ~= 2.6',
    'tensorflow_estimator ~= 2.6',
    'keras ~= 2.6',
    'tensorflow-io-gcs-filesystem >= 0.20.0',
]
```

**Comment by matthewfeickert** [comment]
Similar problematic behavior is unfortunately continuing with https://github.com/tensorflow/probability/issues/1723

**Reply by mihaimaruseac** [reply]
Unfortunately, it's very hard to do a TF 3.0, and APIs need to be deprecated/replaced/changed. See https://github.com/tensorflow/community/pull/450 for the current mechanism to achieve these goals

**Comment by matthewfeickert** [comment]
[TensorFlow `v2.14.0`](https://github.com/tensorflow/tensorflow/releases/tag/v2.14.0) breaks TensorFlow Probability at import. https://github.com/tensorflow/probability/issues/1752

**Reply by matthewfeickert** [reply]
c.f. PR https://github.com/scikit-hep/pyhf/pull/2342

**Reply by jonas-eschle** [reply]
Ah yes, this can happen, usually the TFP versions are released against a specific TF/JAX version, i.e. 0.21 is against TF 2.13, and the release usually happens a couple of days, weeks after it.

What is helpful is to test in this case against the nightlies, they have them in "tf-nightly" or "tfp-nigthly", seems like "tensorflow2.14" against "tfp-nightly" works

**Reply by matthewfeickert** [reply]
Yes, I'm aware. Though, no, it is not helpful to test against the `*-nightly` wheels. The HEAD of TFP is not a rc and what matters is testing in user space.

**Reply by jonas-eschle** [reply]
Yeah, fully agree, I meant it "may" helps to preview a bit, but you're of course right!
I tend to therefore restrict the TFP version not to break previous releases. Conservative, maybe, but the code runs for sure also in the future... technically, TFP should do that but they don't, AFAIU, because of different TF versions, i.e. TF-core, TF-macos, TF-gpu (some are not used anymore, but still)

**Reply by alexander-held** [reply]
Now that the temporary restriction of v2.14.0 is removed again in `pyhf` via #2344, I was looking forward to this resolving a dependency clash I observed via `typing_extensions`. Unfortunately, `tensorflow-probability` introduces another clash: https://github.com/tensorflow/probability/issues/1753 (my specific case is https://github.com/scikit-hep/cabinetry/issues/428).

**Reply by matthewfeickert** [reply]
:disappointed:

**Comment by matthewfeickert** [comment]
c.f. https://github.com/tensorflow/probability/issues/1795 and PR https://github.com/scikit-hep/pyhf/pull/2452.

A reminder that `pyhf` is dropping support for `tensorflow` in `pyhf` `v0.9.0` given that `tensorflow` is too brittle of an ecosystem to use in library development.

**Reply by mxfactorial** [reply]
dependency pinning isnt the only antipattern in tensorflow

#### autodiff in tensorflow
- you build a computational graph (chain of operations)
- every op (multiply add sin cos etc) gets saved
- during backprop you walk backward through that graph
- apply chain rule at each step
- it needs memory to store all intermediate values
- it scales with the size and complexity of your model

#### autodiff in https://crates.io/crates/geonum
geonum doesnt build a computational graph  
it doesnt need to because  
- differentiation is a rotation
- you differentiate by adding π/2 to the angle: [r, θ]' = [r, θ + π/2], ie, the derivative of a number is just another number rotated 90 degrees

this happens because sin and cos are baked directly into the angle encoding:
- derivative of sin(x) is cos(x)
- cos(x) = sin(x + π/2)

so when you rotate the angle by 90 degrees you are implicitly performing differentiation, meaning:
- no graph to build
- no graph to store
- no graph to walk
- no chain rule to manually apply
- no memory explosion
- no slowdown at large model sizes
- autodiff becomes o1

#### practical example
if you want to compute the gradient of a loss function  
you don't need backward()  

you just:

1. rotate the prediction angle +π/2
2. rotate the input angle +π/2
3. rotate the output angle +π/2
4. perform your angle algebra

and youre done

loss minimization becomes:
1. compare angles
2. rotate toward minimization

```rs
use geonum::*;

fn gradient_step_scaled(pred: Geonum, truth: Geonum, base_learning_rate: f64) -> Geonum {
    let pred_rotated = Geonum { angle: pred.angle + std::f64::consts::FRAC_PI_2, ..pred };
    let truth_rotated = Geonum { angle: truth.angle + std::f64::consts::FRAC_PI_2, ..truth };

    let angle_error = pred_rotated.angle - truth_rotated.angle;
    let scaling = angle_error.cos().abs(); // scale learning rate by local cosine similarity
    let effective_lr = base_learning_rate * scaling;

    Geonum {
        angle: pred.angle - effective_lr * angle_error,
        ..pred
    }
}

fn main() {
    let mut pred = Geonum { length: 1.0, angle: 0.0, blade: 0 };
    let truth = Geonum { length: 1.0, angle: 1.0, blade: 0 };

    for step in 0..200 {
        pred = gradient_step_scaled(pred, truth, 0.1);

        if step % 10 == 0 {
            println!("step {}: angle error = {}", step, (pred.angle - truth.angle).abs());
        }

        if (pred.angle - truth.angle).abs() < 0.001 {
            println!("converged at step {}", step);
            break;
        }
    }

    assert!((pred.angle - truth.angle).abs() < 0.001);
}

```

in tensor world:

autodiff is often the main bottleneck for memory and speed

in geonum:

differentiation is just a shift
complexity is constant o1
scaling is free
million parameter models can be differentiated without growing memory or runtime
you dont need a runtime graph to differentiate anymore, you just need geometry

if youre open to leaving O(k^n) tensors, O(n^3) matrices and O(n depth) autodiff behind with the python runtime, try it out: https://github.com/mxfactorial/geonum?tab=readme-ov-file#learn-with-ai

**Reply by mihaimaruseac** [reply]
That doesn't generalize to any differentiable loss function, though.

**Reply by mxfactorial** [reply]
it does generalize

because differentiable loss is computed as the geometric distance in angle space and not as a tree of ops

tensorflow builds a graph

so you need to walk it

but geonum rotates an angle and calls it a day because differentiation is an angle shift and not a traversal, [tests/calculus_test.rs](https://github.com/mxfactorial/geonum/blob/develop/tests/calculus_test.rs):

```rs
    // differentiation is simply a pi/2 rotation and the foundation of
    // calculus emerges directly from this geometric structure

    // let v = [[1, 0], [1, pi/2]] # 2d

    // everything can be a 1d "derivative" or projection of the base 2d v
    // so long as the difference between their angles is pi/2 and they
    // follow the "lengths multiply, angles add" rule

    // v'       = [1, pi/2]  # first derivative (rotate v by pi/2)
    // v''      = [1, pi]    # second derivative (rotate v' by pi/2) = -v
    // v'''     = [1, 3pi/2] # third derivative (rotate v'' by pi/2) = -v'
    // v''''    = [1, 2pi]   # fourth derivative (rotate v''' by pi/2) = v
    // v'''''   = [1, 5pi/2] # fifth derivative (rotate v'''' by pi/2) = v'
    // v''''''  = [1, 3pi]   # sixth derivative (rotate v''''' by pi/2) = -v
    // v''''''' = [1, 7pi/2] # seventh derivative (rotate v'''''' by pi/2) = -v'

    // this geometric space enables continuous rotation as an
    // incrementing pi/2 angle, which is the essence of differentiation,
    // and sets the period of the "derive" function to 4
```

so any loss that can be expressed as a difference in angle or a function of that difference is natively differentiable at o1

no need to build or walk a graph—because the structure of change is already encoded in the geometric number itself

if you still need a graph to differentiate you arent modeling change, you are cataloging syntax and chasing derivatives across operations you could replace with a single rotation

bottom line: the accumulated change in your angle component **is the gradient**

read [machine_learning_test.rs](https://github.com/mxfactorial/geonum/blob/develop/tests/machine_learning_test.rs)

run it: https://github.com/mxfactorial/geonum?tab=readme-ov-file#learn-with-ai

extend it

watch as your code suddenly enables your cpu and memory to execute the `its_a_million_body_simulation()` in milliseconds, [tests/astrophysics_test.rs](https://github.com/mxfactorial/geonum/blob/develop/tests/astrophysics_test.rs):

```sh
cargo test --test astrophysics_test its_a_million_body_simulation -- --show-output
    Finished `test` profile [unoptimized + debuginfo] target(s) in 0.05s
     Running tests/astrophysics_test.rs (target/debug/deps/astrophysics_test-d6eedeaab46743d8)

running 1 test
test its_a_million_body_simulation ... ok

successes:

---- its_a_million_body_simulation stdout ----
Expected velocity: 3653573.593072952, Actual velocity: 3653573.593072952, Ratio: 1
Million-body simulation analysis:
  Initialization time (10 bodies): 3.887µs
  Estimated time for 1M bodies: 388.7ms
  Traditional operations per timestep: 10¹²
  Geonum operations per timestep: 10⁶
  Theoretical speedup: 10⁶×
  Critical advantage: O(1) operations regardless of dimension
  This enables physics that would be impossible to simulate otherwise:
  - Relativistic corrections for each body
  - Quantum effects in stellar evolution
  - Dark matter/energy interactions
  - Multi-scale physics from stars to superclusters


successes:
    its_a_million_body_simulation

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 4 filtered out; finished in 0.00s
```

**Reply by jonas-eschle** [reply]
I must have missed this in calculus class.

>  the derivative of a number is just another number

yes, agree.

>  rotated 90 degrees

Could you explain how to rotate a number?

> because differentiable loss is computed as the geometric distance in angle space and not as a tree of ops

I don't think so. Can you explain how a likelihood measures the geometric distance in space? Do you have any reference for the math?

I have the impression that the library _may_ fits certain use-cases when you deal with rotations, but it doesn't seem to fit the general use-case. I must wonder: how did you come accross this thread?

**Comment by mxfactorial** [comment]
> Could you explain how to rotate a number?

yes, its just adding pi/2 to the angle component: https://github.com/mxfactorial/geonum?tab=readme-ov-file#eli5

> Can you explain how a likelihood measures the geometric distance in space?

in geonum the difference between predicted and observed values is an angular difference  

you can compute loss as a function of that angle, then update by rotating the prediction toward the target

requiring an angle component in all numbers doesnt restrict it to rotation problems 

it removes the need for traversal-based autodiff entirely by preserving geometric meaning in the number itself

> I must wonder: how did you come across this thread?

youre tensor users

"tensor" is math speak for "i dont know how to set a metric for transforming vectors so ill just punt a k^n vector component permutation problem to computers and lose them to testing for orthogonality in that blackhole instead of simulating million body problems"

you strike as people who want to simulate million body problems (physics)

and not pursuing tensors as some kind of math religion where punting metric design to brute force is a sacrificial rite

> I don't think so.

weve volleyed enough

believe in science and test what you think: https://github.com/mxfactorial/geonum?tab=readme-ov-file#learn-with-ai

**Reply by mihaimaruseac** [reply]
Nowhere in your project have you defined addition/subtraction for your numbers.

But even assuming it's lengths add if they're aligned (and don't care for now for what happens if the angles are different), you'd get into a contradiction. Consider $f(x) = 2x$ and let's compute the derivative. Using the limit formulation and denoting your numbers as $[length, angle]$:

$$f'(x) = lim_{t \rightarrow 0} \frac{[2(x+t), 0] - [2x, 0]}{[t, 0]} = lim_{t \rightarrow 0} \frac{[2t, 0]}{[t, 0]}$$

Using [your code](https://github.com/mxfactorial/geonum/blob/3647def2aa0e7f0b24f05594125df07411a9e22a/src/geonum_mod.rs#L531-L556) this becomes

$$f'(x) = lim_{t \rightarrow 0} [2t, 0][1/t, 0]$$

Multiplying (as you said, angles adds) $f'(x) = [2, 0]$

But you just claim that a derivative is just rotating by half pi. So $f'(x) = [f(x), pi/2]$.

Please check your math.

**Reply by mxfactorial** [reply]
please file your own bugs going forward: https://github.com/mxfactorial/geonum/issues/46

youre also welcome to knock out https://github.com/mxfactorial/geonum/issues/30 if it interests you

**Reply by matthewfeickert** [reply]
This is off topic. I'm going to block further noise.
